### PR TITLE
Allowing to select what to extract from the subsetted Spectra

### PR DESCRIPTION
**Pull request by jorainer**
This PR adds code to enable to specify what to extract from the filtered `Spectra` object.

### CODE REVIEW

File: R/massql-parser.R [hunk 1]
```diff
@@ -52,12 +52,71 @@ NULL
 }
 
 .validate_what <- function(x) {
-    if (any(is.na(x)))
+    if (any(is.na(x)) || !length(x))
```
Comment (sgibb) [inline R/massql-parser.R:55]:
```suggestion
    if (anyNA(x) || !length(x))
```

**Review by sgibb — CHANGES_REQUESTED** [review]
Sorry, but IMHO you wrote too many functions :wink:

File: R/massql-parser.R [hunk 2]
```diff
@@ -52,12 +52,71 @@ NULL
...
+    ## identify the function that should be applied.
+    idx <- grep(tolower(what), names(.WHAT_FUNCTIONS), fixed = TRUE)
+    if (!length(idx))
+        stop("Specified type of data '", what, "' is not supported.",
+             call. = FALSE)
+    if (length(idx) > 1L)
+        stop("Specified type of data '", what, "' is ambiguous.", call. = FALSE)
+    fun <- .WHAT_FUNCTIONS[idx]
+    do.call(fun, list(x, what = what))
+}
```
[the remark this answers is not in the document]
Comment (jorainer) [reply R/massql-parser.R:86]: OK. Fine for me.

[the remark this answers is not in the document]
Comment (jorainer) [reply R/massql-parser.R:86]: Actually, I will have to re-write this part. In the end we need to support: `*`, `MS1DATA`, `MS2DATA`, `scaninfo(MS1DATA)`, `scansum(MS1DATA)` etc. So, for now I will make a distinction for MS1DATA and MS2DATA with the additional functions (scaninfo etc) being supported later.

[the remark this answers is not in the document]
Comment (jorainer) [reply R/massql-parser.R]: I've now re-implemented the function to allow also later parsing of functions to extract data subsets (e.g. `scaninfo()` or `scansum()`).

File: NAMESPACE [hunk 3]
```diff
@@ -8,6 +8,17 @@ importFrom(Spectra,containsMz)
 importFrom(Spectra,filterPolarity)
 importFrom(Spectra,filterPrecursorCharge)
 importFrom(Spectra,filterRt)
+importMethodsFrom(Spectra,"*,")
+importMethodsFrom(Spectra,"ms1data,")
+importMethodsFrom(Spectra,Define)
+importMethodsFrom(Spectra,Supported)
```
Comment (sgibb) [inline NAMESPACE:14]: See `roxygen2` documentation for `.what_data` below.

File: NAMESPACE [hunk 4]
```diff
@@ -8,6 +8,17 @@ importFrom(Spectra,containsMz)
...
 importMethodsFrom(Spectra,acquisitionNum)
+importMethodsFrom(Spectra,are)
```
Comment (sgibb) [inline NAMESPACE:16]: See `roxygen2` documentation for `.what_data` below.

File: NAMESPACE [hunk 5]
```diff
@@ -8,6 +8,17 @@ importFrom(Spectra,containsMz)
...
 importMethodsFrom(Spectra,containsNeutralLoss)
+importMethodsFrom(Spectra,data)
+importMethodsFrom(Spectra,extract.)
```
Comment (sgibb) [inline NAMESPACE:19]: See `roxygen2` documentation for `.what_data` below.

File: NAMESPACE [hunk 6]
```diff
@@ -8,6 +8,17 @@ importFrom(Spectra,containsMz)
...
+importMethodsFrom(Spectra,filterMsLevel)
 importMethodsFrom(Spectra,filterPrecursorMz)
+importMethodsFrom(Spectra,ms2data.)
+importMethodsFrom(Spectra,to)
+importMethodsFrom(Spectra,which)
```
Comment (sgibb) [inline NAMESPACE:24]: See `roxygen2` documentation for `.what_data` below.

**Review by sgibb — APPROVED** [review]
I am happy with your code but there is something weird with the `roxygen2` tags.

After fixing it you could merge it.

File: R/massql-parser.R (hunk 2 above)
[the remark this answers is not in the document]
Comment (sgibb) [reply R/massql-parser.R]: Sorry, you see I don't know anything about the syntax of MassQL :smile: (just assuming its like SQL but recognized in an earlier PR it is not (e.g. combination of AND/OR))

[the remark this answers is not in the document]
Comment (jorainer) [reply R/massql-parser.R]: I know - I'm also still thinking in SQL terms. Not so sure if it was a good decision to have a query language similar, but still different than SQL...

**Comment by jorainer** [comment]
Thanks @sgibb for the review. I fixed now the roxygen issue - seems roxygen doesn't like if there is text/description next to `@import`. Waiting for review from @andreavicini before merging.

File: R/Spectra.R [hunk 7]
```diff
@@ -1,30 +1,100 @@
...
+#'
+#' The `<type of data>` allows to define which data should be extracted from
+#' the selected spectra. MassQL defines *type of data* being `MS1DATA` or
+#' `MS2DATA` to retrieve MS1 or MS2 scans. In addition, functions can be applied
+#' to these to modify (e.g. sum) the data. `SpectraQL` supports:
+#'
+#' - `*`: select all data and return the data subset as a [Spectra()] object.
+#' - `MS1DATA`: return a [Spectra()] with all MS1 scans from the selected
+#'   spectra.
+#' - `MS1DATA`: return a [Spectra()] with all MS2 scans from the selected
```
Comment (andreavicini) [inline R/Spectra.R:30]: Maybe `MS2DATA`?

**Review by andreavicini — APPROVED** [review]
The changes seem very nice to me, thanks!

**Pull request merged**